sbtc/src/idpack/codec/decoder.rs

```diff
@@ -0,0 +1,352 @@
+    ///
+    /// * `Ok(Segments)` - Successfully decoded segments
+    /// * `Err(DecodeError)` - If any error occurs during decoding
+    ///
+    /// ## Implementation Notes
+    /// - Returns empty segments for empty input (valid edge case)
+    /// - Ensures complete consumption of input bytes
+    /// - Maintains offset ordering constraints between segments
+    fn decode(bytes: &[u8]) -> Result<Self, DecodeError> {
+        let mut segments = Segments::default();
+
+        if bytes.is_empty() {
+            return Ok(segments);
+        }
+
+        let mut cursor = Cursor::new(bytes);
+
+        let mut prev_max_value = 0; // Tracks previous segment's max value for delta decoding
+
+        // Process segments until we've consumed all input bytes
+        while cursor.position() < bytes.len() as u64 {
+            // Read next segment with position-aware offset handling
+            let segment = read_segment_into(&mut cursor, prev_max_value)?;
+
+            // Update state for next segment
+            prev_max_value = segment.max(); // Use max value for delta encoding
+            segments.try_push(segment)?;
+        }
+
+        // Ensure we've consumed exactly the right amount of data
+        if cursor.position() as usize != bytes.len() {
+            return Err(DecodeError::UnexpectedEndOfData);
+        }
+
+        Ok(segments)
+    }
+}
+
+/// Reads a single segment from the cursor with delta-offset optimization.
+///
+/// This function decodes a single segment from the current cursor position,
+/// applying delta-offset decoding relative to the provided previous maximum
+/// value. It handles the complete segment decoding process:
+/// 1. Reading and decoding the offset value
+/// 2. Reading the payload length and validating allocation size
+/// 3. Reading and processing the bitmap payload
+///
+/// ## Parameters
+/// * `cursor` - Mutable cursor positioned at the start of a segment
+/// * `prev_max_value` - Previous segment's maximum value (0 for first segment)
+///
+/// ## Returns
+/// * `Ok(Segment)` - Successfully decoded segment
+/// * `Err(DecodeError)` - If any error occurs during decoding
+///
+/// ## Errors
+/// * `ArithmeticOverflow` - If offset calculation would overflow
+/// * `ByteAllocationLimit` - If payload length exceeds allocation limits
+/// * `IO` - For any I/O errors during reading
+/// * `Leb128Error` - For LEB128 decoding errors
+pub fn read_segment_into(
+    cursor: &mut Cursor<&[u8]>,
+    prev_max_value: u64,
+) -> Result<Segment, DecodeError> {
+    // Read LEB128-encoded offset value
+    let offset = cursor
+        .read_leb128()?
+        .checked_add(prev_max_value)
+        .ok_or(DecodeError::ArithmeticOverflow)?;
+
+    // Read the payload length
+    let payload_length = cursor.read_leb128()?;
+
+    // Safety check to prevent excessive allocation
+    if payload_length > crate::idpack::ALLOC_BYTES_LIMIT as u64 {
+        return Err(DecodeError::ByteAllocationLimit(payload_length));
+    }
```

Comment:
This is needed because of the unittests/proptests or because in our `OP_RETURN` is only ~80 bytes (or both)? And if it's because of the `OP_RETURN` standardness limits, why not make the limit 80 bytes instead of 1 MB?

Reply:
Some sort of guard is needed so that a malformed length header (malicious or otherwise) can't OOM the signer with e.g. a `u64::MAX` byte count. So the value has no real semantic relevance, it's just a safety net and is hopefully never actually used (outside of garbage-decode tests).

Reply:
I see, makes sense. But I think I have a similar question: why not make the limit 80 bytes instead of 1 MB?

Reply:
Largely just because having a larger limit lets us run tests against larger data which is a nice comfort knowing that no bugs lie just beyond 80 bytes or something (for whatever reason). Really this should be a decoding option instead of a constant, but I didn't want to make any more changes now.

But also because I'm generally against tight-coupling components, setting it to 80 would explicitly tie this module to OP_RETURNs current (but not protocol-enforced) byte limits.

Reply:
Okay, 👍🏽